src/NuGet.Core/NuGet.Protocol/Providers/ServiceIndexResourceV3Provider.cs

```diff
@@ -156,6 +159,69 @@ public ServiceIndexResourceV3Provider()
                         log.LogMinimal(message);
                         throw;
                     }
+                    catch (HttpRequestException ex) when (retry < maxRetries)
+                    {
+                        HttpStatusCode? statusCode = null;
+#if NETCOREAPP5_0
+                        statusCode = ex.StatusCode;
+#else
+                        // HttpRequestException (until .net 5.0, doesn't have ability get status code from http response), so we parse exception string.
+                        var searchFor = ": 40";
+                        int searchForLength = searchFor.Length;
+                        int statusCodeStart = ex.Message.IndexOf(searchFor, 0, StringComparison.OrdinalIgnoreCase);
```

Comment:
Parsing exception text scares me. During the .NET Core package validation work, I found some code (I don't remember if it was test code or production code) which also parsed exception text and failed because .NET 5 changed the exceptions' text message. Using `#if NETCOREAPP5_0` somewhat mitigates the risk, because newer TFMs will use that codepath which gets a strongly typed StatusCode. But it feels fragile, and I don't know what languages you tested with, particularly Chinese and Japanese, which I'm not so confident will use `: `

I was wrong about HttpStatus being in the `httpSourceResult` in the anonymous method above, but did you validate if the code actually hits that method?  If it does, maybe it's better/safer to add `HttpStatus?` to `HttpSourceResult`.

Otherwise, all exceptions have a property bag. Can we catch the exception in `HttpRetryHandler`, where we have the `HttpResponseMessage` object, and populate the property bag with the http, so that we can retrieve it here?

Reply:
I'll spend a little more time digging into this. 
I've improved the code, but I do understand what you mean, so i will see if your latest suggestion is tenable...

The CLR team resisted doing this feature for a few years, but the only workaround they gave included this technique. Yes, it may not work in every language...but it will fail gracefully - maintaining current behavior.

Reply:
wow, ex.Data["StatusCode"] = statusCode works pretty well.
downside - need to find all call sites that may raise the exception.
but nice. thx.